### PR TITLE
kubernetes based discovery druid extension to run Druid on K8S without Zookeeper

### LINKED ISSUE #9053: [Proposal] Druid discovery extension for Kubernetes
### Motivation

Druid has provisions for writing an extension that could be used for "node discovery" and "leader election". Such extension coupled with "HTTP" based "segment management" and "remote task runner" could be used to obviate the need of operating a Zookeeper cluster for Druid.
However, no such extension exists and above is just theory.

Proposal is to write an extension based on APIs exposed by Kubernetes API server (which itself is backed by [etcd](https://github.com/etcd-io/etcd) ) which has sufficient primitives to be able to do  "node discovery" and "leader election" .

### Proposed changes

A new extension, "druid-kubernetes-extensions" , would be added with implementations of various discovery related interfaces such as "DruidNodeDiscoveryProvider", "DruidLeaderSelector", "DruidNodeAnnouncer" etc.
Additionally, since this is first such extension, there might be some changes needed in core as well to enable writing the extension.

### Rationale

etcd based extension could also be written, but in many Kubernetes deployments, access to etcd is guarded and users of K8s clusters are not expected to use it. So for users(I happen to be one of those) deploying Druid on kubernetes, it is required to not depend directly on etcd. So, I decided not to pursue that.
it is also possible to have Druid implement one of the consensus algorithms, but that is a lot more work and is very error prone. Consensus algorithms are notoriously hard to implement. In the current use case, that is not strictly needed.

More importantly, on most cloud providers, K8S control plane comes with a fixed cost so using it instead of zookeeper lets you remove 3 zk pods for free and adds to cloud cost savings.

### Operational impact

None

### Test plan (optional)

I would be testing the extension on my own Druid clusters deployed in Kubernetes.

### PR DESCRIPTION
**Pull request by himanshug**
Fixes #9053 

### Description

Please read #9053 first and maybe https://groups.google.com/g/druid-development/c/tWnwPyL0Vk4/m/2uLwqgQiAAAJ?pli=1 for more background on HTTP based segment and task management that were introduced before.

This patch has been tested to successfully run a small Druid Test cluster with K8S and without Zookeeper. 

Most of the code introduced here goes in a new extension. At a high level, it introduces a new Druid Kubernetes Extension that implements 3 druid discovery and leader election related interfaces... "DruidNodeDiscoveryProvider", "DruidLeaderSelector", "DruidNodeAnnouncer" and provides necessary plumbing to use those when configured.

Internally, it uses https://github.com/kubernetes-client/java for all the kubernetes interactions and https://github.com/kubernetes-client/java/tree/master/extended/src/main/java/io/kubernetes/client/extended/leaderelection for the leader election.

See `kubernetes.md` for how to use it. It can support multiple Druid clusters running on same K8S cluster [in same namespace].

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [x] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [x] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [x] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests. See https://github.com/apache/druid/issues/10542
- [x] been tested in a test Druid cluster.

<hr>

Note: This patch also adds "Bouncy Castle License" being same as "MIT License" for license checking purposes in `distribution/bin/check-licenses.py` . See http://www.bouncycastle.org/licence.html

### CODE REVIEW

**Comment by himanshug** [comment]
This PR is no longer WIP. Build is passing, there is good amount of test coverage and I have also tested it successfully on small test clusters running on kubernetes without a zookeeper cluster at all. This PR is ready to be merged.

While I am sure caveats would pop up when this gets used in long running large clusters, this is a good starting point to be released as an experimental feature in next Druid release and noted as such in the docs introduced.

sidenote:
if someone has experimented with running k8s clusters in travis builds, please feel free to work/comment on https://github.com/apache/druid/issues/10542 as that would set the stage of testing the code here on real k8s cluster in the travis builds.

File: extensions-core/kubernetes-extensions/src/main/java/org/apache/druid/k8s/discovery/DiscoveryDruidNodeList.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import java.util.Map;
+
+public class DiscoveryDruidNodeList
+{
+  private final String resourceVersion;
+  private final Map<String, DiscoveryDruidNode> druidNodes;
+
+  public DiscoveryDruidNodeList(
+      String resourceVersion,
+      Map<String, DiscoveryDruidNode> druidNodes
```
Comment (nishantmonu51) [inline extensions-core/kubernetes-extensions/src/main/java/org/apache/druid/k8s/discovery/DiscoveryDruidNodeList.java:35]: nit: Add @Nullable annotations.

**Review by nishantmonu51 — APPROVED** [review]
LGTM, 👍 
minor nit: would be great if we can also add @Nullable annotations

File: extensions-core/kubernetes-extensions/src/main/java/org/apache/druid/k8s/discovery/DiscoveryDruidNodeList.java (hunk 1 above)
Comment (himanshug) [reply extensions-core/kubernetes-extensions/src/main/java/org/apache/druid/k8s/discovery/DiscoveryDruidNodeList.java]: added, thanks

**Comment by harinirajendran** [comment]
@himanshug @gianm Do we know of any big production clusters using this feature? how does the migration from zk to k8s based discovery work without downtime?

**Pull request merged**